### ISSUE TITLE
Attaching to a hotspot VM in container may fail in `start_attach_mechanism`

### ISSUE THREAD

**Issue by rvql** [jattach/jattach#77]
Hi,

The problem occurred in a multithreaded program.

According to `setns` manpage, setting user namespace in a multithreaded process is forbidden, jattach cannot enter mount namespace in the program, in which case it should use `tmp_path` (`/proc/$pid/root/tmp`) to create the trigger file `.attach_pid$nspid` instead.

The function `start_attach_mechanism` in `jattach_hostspot.c` tries to create the trigger in `/proc/$nspid/cwd`. If `setns` failed for mount namespace, this path will be in current mount namespace. If the OS does not have a process with PID equal to `$nspid`, the function will fail and fall back to creating the trigger file in `tmp_path`, which should be the right behavior. However, if the OS does have a process with PID `$nspid`, the trigger file will be created successfully, but in a wrong location, and it will not trigger hotspot attach listener.

A possible fix would be to avoid using $nspid path when `setns` for mount namespace fails.

Thanks,
Jie

**Comment by apangin** [comment]
> According to setns manpage, setting user namespace in a multithreaded process is forbidden

How does this relate to the issue? jattach is not a multithreaded process.

**Comment by rvql** [comment]
> > According to setns manpage, setting user namespace in a multithreaded process is forbidden
> 
> How does this relate to the issue? jattach is not a multithreaded process.

The issue occurred when I try to use `libjattach.so` in my multithreaded program.

Multithreading is one of the ways to make setting mount namespace fail, so `jattach` may also suffer from this issue, creating the trigger file at a wrong location, thus unable to attach to JVM.

**Comment by apangin** [comment]
Since jattach does irreversible changes to the current process, the right way to use it as a library is to fork the current process and invoke jattach in a child. See https://github.com/async-profiler/async-profiler/blob/809a19ce8f068aa6ea47699a8917314366bbd492/src/launcher/main.cpp#L355-L377

**Comment by rvql** [comment]
Thanks for the advise. I'm closing the issue if this is not a problem.

**Comment by apangin** [comment]
Yes, this should not be a problem when using jattach in a forked process.
Anyways, your analysis is valid, so I've fixed the issue just in case. Thanks.
